Objects/stringlib/fastsearch.h

```diff
@@ -324,222 +359,402 @@ typedef struct STRINGLIB(_pre) {
+    else {
+        Py_ssize_t cut_idx = m - cut;
+        is_periodic = (0 == memcmp(p + cut_idx, p + cut_idx - period,
+                                   cut * STRINGLIB_SIZEOF_CHAR));
+    }
+    if (is_periodic) {
+        assert(cut <= m/2);
+        assert(cut < period);
+        LOG("Needle is periodic.\n");
     }
+    else {
+        // A lower bound on the period
+        // CLARIFY> An upper bound?
+        period = Py_MAX(cut, m - cut) + 1;
+        LOG("Needle is not periodic.\n");
+    }
+    pw->cut = cut;
+    pw->period = period;
+    pw->is_periodic = is_periodic;
+
+    LOG("Cut: %ld & Period: %ld\n", cut, period);
+    LOG("split: ");
+    LOG_STRING(p, cut);
+    LOG(" + ");
+    LOG_STRING(p + cut, m - cut);
+    LOG("\n");
 }
 
+
 static Py_ssize_t
-STRINGLIB(_two_way)(const STRINGLIB_CHAR *haystack, Py_ssize_t len_haystack,
-                    STRINGLIB(prework) *p)
+STRINGLIB(_two_way)(const STRINGLIB_CHAR *s, Py_ssize_t n,
+                    Py_ssize_t maxcount, int mode,
+                    STRINGLIB(prework) *pw, int direction)
 {
     // Crochemore and Perrin's (1991) Two-Way algorithm.
     // See http://www-igm.univ-mlv.fr/~lecroq/string/node26.html#SECTION00260
-    const Py_ssize_t len_needle = p->len_needle;
-    const Py_ssize_t cut = p->cut;
-    Py_ssize_t period = p->period;
-    const STRINGLIB_CHAR *const needle = p->needle;
-    const STRINGLIB_CHAR *window_last = haystack + len_needle - 1;
-    const STRINGLIB_CHAR *const haystack_end = haystack + len_haystack;
-    SHIFT_TYPE *table = p->table;
-    const STRINGLIB_CHAR *window;
-    LOG("===== Two-way: \"%s\" in \"%s\". =====\n", needle, haystack);
-
-    Py_ssize_t gap = p->gap;
-    Py_ssize_t gap_jump_end = Py_MIN(len_needle, cut + gap);
-    if (p->is_periodic) {
-        LOG("Needle is periodic.\n");
-        Py_ssize_t memory = 0;
-      periodicwindowloop:
-        while (window_last < haystack_end) {
+    if (mode == FAST_COUNT) {
+        LOG("Two-way Count.\n");
+    }
+    else {
+        LOG("Two-way Find.\n");
+    }
+    LOG("haystack: "); LOG_STRING(s, n); LOG("\n");
+    LOG("needle  : "); LOG_STRING(pw->p, pw->m); LOG("\n");
+    int dir = direction < 0 ? -1 : 1;
+    int reversed = dir < 0;
+
+    // Prepare
+    const STRINGLIB_CHAR *const p = pw->p;
+    const Py_ssize_t m = pw->m;
+    SHIFT_TYPE *table = pw->table;
+    const Py_ssize_t gap = pw->gap;
+    const Py_ssize_t cut = pw->cut;
+    int is_periodic = pw->is_periodic;
+    Py_ssize_t period = is_periodic ? pw->period : Py_MAX(gap, pw->period);
+    Py_ssize_t gap_jump_end = Py_MIN(m, cut + gap);
+
+    // Direction Independent
+    const Py_ssize_t w = n - m;
+    const Py_ssize_t m_m1 = m - 1;
+
+    // Direction Dependent
+    const Py_ssize_t p_stt = dir == 1 ? 0 : m - 1;
+    const Py_ssize_t s_stt = dir == 1 ? 0 : n - 1;
+    const Py_ssize_t p_end = dir == 1 ? m - 1 : 0;
+    const Py_ssize_t cut_idx = reversed ? m - cut : cut;
```

Comment:
For what is this supposed to be used?

Reply:
These are the tricks I used to make the same loop run the algorithm in both forward and reverse directions. Benefits of this are:
a) Less code.
b) Easier to make sure both directions are following the same logic.

It took me a bit of time to come up with standardisation, so it can take a bit of time to get your head round it, but once you do it all makes sense and in my opinion ends up being much simpler than having to work with 2 different blocks of code, while having to keep them in sync.

Reply:
No, I mean for what is `cut_idx` supposed to be used. gcc says it's unused:

```none
unused variable ‘cut_idx’ [-Wunused-variable]
```

It's worth looking at the compiler warnings in the files tab.

Reply:
I defined index names as such:
```
lst = [0, 1, 2, 3]   # forsuch list
begin = -1
start = 0
end = 3
stop = 4
```
For reverse direction, equivalents would be:
```
begin = 4
start = 3
end = 0
stop = -1
```
These definitions make 2 problems symmetrical. Everything else flows out of this nicely.

Reply:
Are you still going to do something with it? Otherwise you should remove it:

```suggestion
```

Reply:
You can resolve this now.